core/src/main/java/org/springframework/security/converter/RsaKeyConverters.java

```diff
@@ -103,22 +115,49 @@ private RsaKeyConverters() {
 		KeyFactory keyFactory = rsaFactory();
 		return (source) -> {
 			List<String> lines = readAllLines(source);
-			Assert.isTrue(!lines.isEmpty() && lines.get(0).startsWith(X509_PEM_HEADER),
-					"Key is not in PEM-encoded X.509 format, please check that the header begins with -----"
-							+ X509_PEM_HEADER + "-----");
+			Assert.isTrue(
+					!lines.isEmpty()
+							&& (lines.get(0).startsWith(X509_PEM_HEADER) || lines.get(0).startsWith(X509_CERT_HEADER)),
+					"Key is not in PEM-encoded X.509 format or a valid X.509 certificate, please check that the header begins with -----"
+							+ X509_PEM_HEADER + "----- or -----" + X509_CERT_HEADER + "-----");
+			RSAPublicKey publicKey = null;
 			StringBuilder base64Encoded = new StringBuilder();
 			for (String line : lines) {
-				if (RsaKeyConverters.isNotX509Wrapper(line)) {
+				if (RsaKeyConverters.isNotX509PemWrapper(line) && isNotX509CertificateWrapper(line)) {
 					base64Encoded.append(line);
 				}
 			}
 			byte[] x509 = Base64.getDecoder().decode(base64Encoded.toString());
-			try {
-				return (RSAPublicKey) keyFactory.generatePublic(new X509EncodedKeySpec(x509));
+			if (lines.get(0).startsWith(X509_PEM_HEADER)) {
+				try {
+					publicKey = (RSAPublicKey) keyFactory.generatePublic(new X509EncodedKeySpec(x509));
```

Comment:
Please change this to return the public key as this is more typical of Spring Security code. Subsequently, please also change the `else if` below to an `if`.